doc/error-handling/error_handling_design_spec.md

```diff
@@ -0,0 +1,429 @@
+      * [3.4.3 Error DB](#343-Error-DB)
+        * [3.4.3.1 Error Tables](#3431-Error-Tables)
+        * [3.4.3.2 Error DB Schemas](#3432-Error-DB-Schemas)
+    * [3.5 CLI](#35-CLI)
+      * [3.5.1 Data Models](#351-Data-Models)
+      * [3.5.2 Config commands](#352-Config-commands)
+      * [3.5.3 Show commands](#353-Show-commands)
+      * [3.5.4 Clear commands](#354-Clear-commands)
+  * [4 Flow Diagrams](#4-Flows-diagrams)
+    * [4.1 Route Add Failure](#41-Route-Add-Failure)
+  * [5 Serviceability and Debug](#5-Serviceability-and-Debug)
+  * [6 Warm Boot Support](#6-Warm-Boot-Support)
+  * [7 Scalability](#7-Scalability)
+  * [8 Unit Tests](#8-Unit-Tests)
+  * [9 Unsupported features](#9-Unsupported-features)
+
+# Revision
+
+| Rev |     Date    |       Author       | Change Description          |
+|:---:|:-----------:|:-------------------------|:----------------------|
+| 0.1 |  05/6/2019  | Siva Mukka, Santosh Doke | Initial version       |
+
+
+# About this Manual
+This document describes the design details of error handling framework in SONiC. The framework is responsible for notifying ASIC/SAI programming failures back to applications. It can also be extended to notify other failures in the system.
+
+# Scope
+This document describes the high level design details of error handling framework in SONiC.
+
+# Overview
+SONIC currently does not have a mechanism to propagate ASIC/SAI programming errors back to application. If the SAI CREATE/SET method fails:
+
+- Syncd treats all such failures as fatal irrespective of reason code.
+- Syncd sends a switch shutdown request to Orchestration Agent.
+- Syncd waits indefinitely to restart, switch is no longer manageable/operational.
+
+To address this, a generic framework is introduced to notify errors, instead of sending shutdown request to OrchAgent.
+
+# Definitions/Abbreviation
+###### Table 1: Abbreviations
+| Abbreviation | Description                                                  |
+| ------------ | ------------------------------------------------------------ |
+| SAI          | Switch Abstraction Interface                                 |
+| APP DB       | Application Database                                         |
+| ASIC DB      | ASIC Database                                                |
+| ERROR DB     | Error Database                                               |
+| SWSS         | Switch State Service                                         |
+| Syncd        | Daemon responsible for invoking SAI functions based on ASIC DB updates |
+
+In this document, the term '**object**' refers to an entry in the table OR a specific attribute in the table.
+
+# 1 Requirements Overview
+## 1.1 Functional Requirements
+
+The requirements for error handling framework are:
+
+1.1.1 Provide registration/de-registration mechanism for applications to enable/disable error notifications on a specific table. More than one application can register for notifications on a given table.
+

```

Comment:
Can framework supports applications to register for notifications at attribute level?

Reply:
Notifications can only be enabled per-table. Failure status code is reported per object. For example, port table has multiple objects like MTU, admin state. Notifications can be enabled at Port table level, but not on MTU failures specifically.